docs/proposals/20240930-machine-drain-rules.md

```diff
@@ -0,0 +1,324 @@
+`MachineDrainRule` CRD to make the drain rules configurable per Pod. Additionally, we're proposing annotations that 
+workload cluster admins can add to individual Pods to control their drain behavior.
+
+This would be a huge improvement over the “standard” `kubectl drain` aligned implementation we have today and help to 
+solve a family of issues identified when running Cluster API in production.
+
+## Motivation
+
+While the “standard” `kubectl drain` rules have served us well, new user stories have emerged where a more sophisticated 
+behavior is needed.
+
+### User Stories
+
+**Story 1: Define drain order**
+
+As a cluster admin/user, I would like to define in which order Pods are evicted. For example, I want to configure that 
+Portworx Pods are evicted after all other Pods ([CAPI-11024](https://github.com/kubernetes-sigs/cluster-api/issues/11024)).
+
+Note:  As of today the only way to address this use case with Cluster API is to implement a pre-drain hook with a custom 
+drain process. Due to the complexity of this solution it is not viable for most users.
+
+**Story 2: Exclude “undrainable” Pods from drain**
+
+As a cluster admin/user, I would like the node drain process to ignore pods that will obviously not drain properly 
+(e.g., because they're tolerating the unschedulable or all taints).
+
+Note: As of today Pods that are tolerating the `node.kubernetes.io/unschedulable:NoSchedule` taint will just get rescheduled 
+to the Node after they have been evicted, and this can lead to race conditions and then prevent the Machine to be deleted.
+
+**Story 3: Exclude Pods that should not be drained from drain**
+
+As a cluster admin/user, I would like to instruct the Node drain process to exclude some Pods, e.g. because I don't 
+care if they're gracefully terminated and they should run until the Node goes offline (e.g. monitoring agents).
+
+**Story 4: Define drain order & exclude Pods from drain without management cluster access**
+
+As a workload cluster admin/user, I want to be able to define drain order of Pods and also exclude Pods from drain 
+(similar to Story 1-3, just without management cluster access).
+
+**Story 5: Apply drain configuration to a set of Machines / Clusters**
+
+As a cluster admin/user, I would like to be able to configure the drain behavior for all Machines of a management 
+cluster or for a subset of these Machines without having to configure each Machine individually. I would also like 
+to be able to change the drain configuration without having to modify all Machine objects.
+
+### Goals
+
+* Allow users to customize how/if Pods are evicted:
+  * Allow to define the order in which Pods are evicted
+  * Allow to define if Pods should be excluded from drain
+* Use the current drain default rules as a fallback if neither `MachineDrainRule` CRs nor drain annotations
+  apply to a Pod (this also ensures compatibility with current behavior)
+* Stop waiting for volume detachment for Pods that are excluded from drain
```

Comment:
Is there any use case we want to point out where this would be specially disruptive before but it will be afterwards. e.g. vsphere?
Also do we want align daemonsSets with this behaviour?

Reply:
> Is there any use case we want to point out where this would be specially disruptive before but it will be afterwards. e.g. vsphere?

Can you please expand on this? I'm not sure what before/ afterwards refer to.

> Also do we want align daemonsSets with this behaviour?

Yup, DaemonSet Pods are always excluded from drain. So this would also apply to them. I'll make that clearer

Reply:
Sorry, I meant that skipping waiting for volume detachment [might be a disruptive operation e.g in vsphere](https://github.com/kubernetes-sigs/cluster-api/pull/4945). Before this proposal you had to explicitly signal it, now it is implicit in skip drain. For daemonSets this is also a behavioural change which has the same implications than above.
So it might be good to point this out, maybe it's already clear for everyone so feel free to ignore :)

Reply:
Got it, makes sense!

@chrischdi Just to provide a data point from vSphere (we're aware it might differ across CSI plugins), can you give a quick tl;dr how the system reacted when you were testing skipping some volumes (including the part when the Volume got attached eventually afterwards)

Reply:
Jep, so first of all: it is only about volumes which are "attachable", which e.g. is the case for vSphere CSI, but is e.g. not for the nfs csi driver (https://github.com/kubernetes-csi/csi-driver-nfs)

Let's assume a pod is ignored during drain on purpose (e.g. because its a pod from a daemonset, or in future a drain rule leads to ignoring it).
This pod refers a CSI Volume (e.g. from vSphere), which is RWO (this means it can only be attached on a single node at the same time / no multi-attachment).

Let's say the machine gets removed and during that the pod does not get evicted (because it can't, its a daemonset) and waiting for the volume to detach also gets skipped, the following happens:

1. The node object gets deleted (either by CAPI, which does this on best effort; or normally via CPI at a later point)
2. After some time the pod also vanishes (must be something in kube-controller-manager doing this)
3. After some time, the `persistentvolume-attach-detach-controller` from `kube-controller-manager` deletes the volumeattachment object, see the following logs):

    ```
    I1001 16:01:29.528559       1 reconciler.go:279] "attacherDetacher.DetachVolume started: this volume is not safe to detach, but maxWaitForUnmountDuration expired, force detaching" logger="persistentvolume-attach-detach-controller" duration="6m0s" node="test-md-0-njk5m-xqg4f-vmw2f" volumeName="kubernetes.io/csi/csi.vsphere.vmware.com^ba0ac90e-26a3-47ac-a365-4a8387638d65"
    I1001 16:01:30.123136       1 operation_generator.go:437] DetachVolume.Detach succeeded for volume "pvc-4d6bb049-5f40-41e9-a658-1130375f0f42" (UniqueName: "kubernetes.io/csi/csi.vsphere.vmware.com^ba0ac90e-26a3-47ac-a365-4a8387638d65") on node "test-md-0-njk5m-xqg4f-vmw2f"
    I1001 16:01:30.171017       1 reconciler.go:360] "attacherDetacher.AttachVolume started" logger="persistentvolume-attach-detach-controller" volumeName="kubernetes.io/csi/csi.vsphere.vmware.com^ba0ac90e-26a3-47ac-a365-4a8387638d65" nodeName="test-md-0-njk5m-xqg4f-2nlmt" scheduledPods=["default/minio-ds-hklpv"]
    I1001 16:01:33.259232       1 operation_generator.go:317] AttachVolume.Attach succeeded for volume "pvc-4d6bb049-5f40-41e9-a658-1130375f0f42" (UniqueName: "kubernetes.io/csi/csi.vsphere.vmware.com^ba0ac90e-26a3-47ac-a365-4a8387638d65") from node "test-md-0-njk5m-xqg4f-2nlmt"
    ```

4. With the VolumeAttachment in deletion, the corresponding csi driver should free up the PVC (e.g. see that it is mounted nowhere anymore, because the VM does not exist anymore) and the VolumeAttachment object is gone
5. Now the disk can be mounted again and used by a different pod.

Notes:
* Only point 4 differs per CSI driver.
* If we would not skip waiting for the volume to detach, we would wait forever, except for reaching the wait timeout, because as long as the pod keeps running the volume will be kept attached
* The proposal would allow a cluster admin to skip pods from being drained. If we don't also ignore their volumes during wait for detach, we would wait for detach forever, while the user's intent was to skip getting rid of this pod :-)

Reply:
I hope this helps to clarify, feel free to ask me anything on this :-)

Reply:
Added a note in the "Changes to wait for volume detach" section. Let me know if that covers it